dot/state/grandpa.go

```diff
@@ -26,42 +37,477 @@ var (
+// that belongs to the same finalized node branch. If we don't find such node we update the scheduled
+// change tree roots with the change nodes that belongs to the same finalized node branch
+func (s *GrandpaState) getApplicableScheduledChange(finalizedHash common.Hash, finalizedNumber uint) (
+	change *pendingChange, err error) {
+
+	var changeNode *pendingChangeNode
+	for _, root := range s.scheduledChangeRoots {
+		if root.change.effectiveNumber() > finalizedNumber {
+			continue
+		}
+
+		changeNodeHash := root.change.announcingHeader.Hash()
+
+		// if the change doesn't have the same hash
+		// neither the finalized header is descendant of the change then skip to the next root
+		if !finalizedHash.Equal(changeNodeHash) {
+			isDescendant, err := s.blockState.IsDescendantOf(changeNodeHash, finalizedHash)
+			if err != nil {
+				return nil, fmt.Errorf("cannot verify ancestry: %w", err)
+			}
+
+			if !isDescendant {
+				continue
+			}
+		}
+
+		// the changes must be applied in order, so we need to check if our finalized header
+		// is ahead of any children, if it is that means some previous change was not applied
+		for _, child := range root.nodes {
+			isDescendant, err := s.blockState.IsDescendantOf(child.change.announcingHeader.Hash(), finalizedHash)
+			if err != nil {
+				return nil, fmt.Errorf("cannot verify ancestry: %w", err)
+			}
+
+			if child.change.announcingHeader.Number <= finalizedNumber && isDescendant {
+				return nil, errUnfinalizedAncestor
+			}
+		}
+
+		changeNode = root
+		break
+	}
+
+	// if there is no change to be applied then we should keep only
+	// the scheduled changes which belongs to the finalized header
+	// otherwise we should update the scheduled roots to be the child
+	// nodes of the applied scheduled change
+	if changeNode == nil {
+		err := s.pruneScheduledChanges(finalizedHash)
+		if err != nil {
+			return nil, fmt.Errorf("cannot prune non-descendant scheduled nodes: %w", err)
+		}
+	} else {
+		change = changeNode.change
+		s.scheduledChangeRoots = make([]*pendingChangeNode, len(changeNode.nodes))
+		copy(s.scheduledChangeRoots, changeNode.nodes)
+	}
+
+	return change, nil
+}
+
+// pruneForcedChanges removes the forced changes from GRANDPA state for any block
+// that is not a descendant of the current finalised block.
+func (s *GrandpaState) pruneForcedChanges(finalizedHash common.Hash) error {
+	onBranchForcedChanges := make([]*pendingChange, 0, len(s.forcedChanges))
+
+	for _, forcedChange := range s.forcedChanges {
+		isDescendant, err := s.blockState.IsDescendantOf(finalizedHash, forcedChange.announcingHeader.Hash())
+		if err != nil {
+			return fmt.Errorf("cannot verify ancestry: %w", err)
+		}
+
+		if isDescendant {
+			onBranchForcedChanges = append(onBranchForcedChanges, forcedChange)
+		}
+	}
+
+	s.forcedChanges = make(orderedPendingChanges, len(onBranchForcedChanges))
+	copy(s.forcedChanges, onBranchForcedChanges)
```

Comment:
no need for the copy, since `onBranchForcedChanges` is not used anywhere else.
```suggestion
	s.forcedChanges = onBranchForcedChanges
```